bpf/tests/tc_nodeport_l3_dev.c

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#include "common.h"
+
+#include <bpf/ctx/skb.h>
+#include <bpf/helpers_skb.h>
+#include "pktgen.h"
+
+#define ETH_HLEN		0
+#define ENABLE_HOST_ROUTING	1
+#define ENABLE_IPV4		1
+
+#define TEST_IP_LOCAL	v4_pod_one
+#define TEST_IP_REMOTE	v4_pod_two
+#define TEST_LXC_ID_LOCAL	233
+
+/* We wanted to tail call handle_policy from bpf_lxc, but at present it's
+ * impossible to #include both bpf_host.c and bpf_lxc.c at the same time.
+ * Therefore, we created a stud, mock_hanle_policy, to simply check if the
+ * our skb reaches there.
+ */
+__section("mock-handle-policy")
+int mock_handle_policy(struct __ctx_buff *ctx __maybe_unused)
+{
+	return TC_ACT_REDIRECT;
+}
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(max_entries, 256);
+	__array(values, int());
+} mock_policy_call_map __section(".maps") = {
+	.values = {
+		[TEST_LXC_ID_LOCAL] = &mock_handle_policy,
+	},
+};
+
+#define tail_call_dynamic mock_tail_call_dynamic
+static __always_inline __maybe_unused void
+mock_tail_call_dynamic(struct __ctx_buff *ctx __maybe_unused,
+		       const void *map __maybe_unused, __u32 slot __maybe_unused)
+{
+	tail_call(ctx, &mock_policy_call_map, slot);
+}
+
+static volatile const __u8 *ep_mac = mac_one;
+static volatile const __u8 *node_mac = mac_two;
+
+#include "bpf_host.c"
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(max_entries, 2);
+	__array(values, int());
+} entry_call_map __section(".maps") = {
+	.values = {
+		[0] = &cil_from_netdev,
+	},
+};
+
+PKTGEN("tc", "l3_to_l2_fast_redirect")
+int l3_to_l2_fast_redirect_pktgen(struct __ctx_buff *ctx)
+{
+	struct pktgen builder;
+	struct ethhdr *l2;
+	struct iphdr *l3;
+	struct tcphdr *l4;
+	void *data;
+
+	/* Init packet builder */
+	pktgen__init(&builder, ctx);
+
+	/* We are building an L3 skb which doesn't have L2 header, so in theory
+	 * we need to set ctx->protocol = bpf_ntohs(ETH_P_IP), but bpf verifier
+	 * doesn't allow us to do so, and kernel also doesn't handle an L3 skb
+	 * properly (see https://elixir.bootlin.com/linux/latest/source/net/bpf/test_run.c#L1156).
```

Comment:
Nit: Switch the link for a permalink version.
```suggestion
	 * properly (see https://elixir.bootlin.com/linux/v6.2.1/source/net/bpf/test_run.c#L1156).
```